rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,396 @@
+
+### The Policy Server is a StatefulSet
+
+The Policy Server will be deployed as a StatefulSet to ensure that each replica has a unique identity.
+This will allow us to identify the status of a policy on a specific Policy Server replica.
+
+### Policy Status
+
+We will use the [status condition](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) convention to report the status of the policy.
+The status condition struct will be extended to include the replica number.
+The approach of extending the status condition struct is already used in the Kubernetes codebase,
+for example in the [Pod conditions](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions).
+
+The `PolicyStatusCondition` struct will have the following fields:
+
+| **Field name**         | **Description**                                                                                     |
+| ---------------------- | --------------------------------------------------------------------------------------------------- |
+| **type**               | Name of the policy condition.                                                                       |
+| **status**             | Indicates whether the condition is applicable, with possible values: "True", "False", or "Unknown". |
+| **reason**             | Machine-readable reason for the condition's last transition, in UpperCamelCase format.              |
+| **message**            | Human-readable message providing details about the last transition.                                 |
+| **replica** (optional) | The replica number reporting the status.                                                            |
+
+The `Policy` CRD and the `PolicyRevision` CRD will be updated to include the status conditions.
+See the [Policy and PolicyRevision conditions](#policy-and-policyrevision-conditions) section for more details on how the status conditions will be propagated.
+
+The following `PolicyConditions` will be observed:
+
+- `Scheduled`: The policy was created and is pending to be loaded.
+- `Initialized`: The policy was successfully pulled, precompiled and validated by the leader.
+- `Ready`: The policy was successfully loaded and is ready to be used. This status will include the replica number.
+
+By using the `reason` field in the status condition, we can provide detailed information in case of an error.
+For instance if the leader fails to initialize a policy because of registry error, the following status condition will be set:
+
+```yaml
+conditions:
+  - type: Initialized
+    status: False
+    reason: PullError
+    message: "Cannot pull the policy from the registry."
+```
+
+### Rollback
+
+If a policy fails to load, the Policy Server will keep running with the previous valid policy.
```

Comment:
I wonder if we can inform the user of this. As it stands, users would need to look at the policy status + generation number, which isn't ergonomic. Maybe we could add a `ReadyButOld` status if the generation is not the last possible.